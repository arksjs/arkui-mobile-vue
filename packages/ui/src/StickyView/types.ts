import type { ShallowRef } from 'vue'
import type { ResetContainer } from '../Sticky/types'

export interface StickyViewItem {
  name: string
  index: number
  title: string
}

export type OnResetItems = (items: StickyViewItem[]) => void

export type OnChange = (name: string, activeIndex: number) => void

export interface StickyViewListRef {
  ref: HTMLElement
}

export interface StickyViewRef {
  scrollTo: (name: string) => void
  scrollToIndex: (index: number) => void
  scrollToOffset: (offset: number) => void
  resetContainer: ResetContainer
}

export interface StickyViewProps {
  modelValue?: string
  containSelector?: HTMLElement
  offsetTop?: number | string
  disabled?: false
}

export interface StickyViewEmits {
  onChange?: OnChange
  onResetItems?: OnResetItems
}

export interface StickyViewItemProps {
  name: string
  title?: string
}

export type {
  OnChange as StickyViewOnChange,
  OnResetItems as StickyViewOnResetItems
}
