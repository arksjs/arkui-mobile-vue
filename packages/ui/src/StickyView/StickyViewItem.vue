<template>
  <div
    class="ta-sticky-view-item"
    :data-name="name"
    :data-title="title"
    ref="root"
  >
    <div class="ta-sticky-view-item_header">{{ title || name }}</div>
    <div class="ta-sticky-view-item_body"><slot></slot></div>
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import { useListItem } from '../hooks/use-list'

export default defineComponent({
  name: 'ta-sticky-view-item',
  props: {
    title: {
      type: String,
      default: ''
    },
    name: {
      type: String,
      required: true
    }
  },
  setup() {
    useListItem('stickyView')

    return {}
  }
})
</script>
