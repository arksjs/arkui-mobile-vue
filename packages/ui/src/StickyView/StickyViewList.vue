<script lang="ts">
import {
  defineComponent,
  h,
  isVNode,
  shallowRef,
  type VNode,
  type Component
} from 'vue'

export default defineComponent({
  name: 'ta-sticky-view-items',
  emits: ['reset-items'],
  setup(_, { slots, emit, expose }) {
    const root = shallowRef<HTMLElement | null>(null)

    expose({
      ref: root
    })

    return () => {
      const children = slots.default?.()

      let items: VNode[] = []
      if (children && isVNode(children[0])) {
        if (Array.isArray(children[0].children)) {
          items = children[0].children
            .map(item => {
              if (isVNode(item) && Array.isArray(item.children)) {
                return item.children.filter(
                  item =>
                    isVNode(item) &&
                    (item.type as Component).name === 'ta-sticky-view-item'
                )
              }

              return []
            })
            .flat() as VNode[]
        }
      }
      emit('reset-items', items)

      return h('div', { class: 'ta-sticky-view_list', ref: root }, children)
    }
  }
})
</script>
