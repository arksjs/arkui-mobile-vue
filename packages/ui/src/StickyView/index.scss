@import '../style/var.scss';

.ta-sticky-view {
  position: relative;

  &.self {
    .ta-sticky-view_content {
      height: 100%;
    }
  }

  &_top {
    position: absolute;
    left: 0;
    top: 0;
    width: 100%;

    .ta-sticky_content {
      overflow: hidden;
    }
  }

  &_fixed {
    width: 100%;
    height: 28px;
    overflow: hidden;
    transform: translate3d(0, 0, 0);

    &-inner {
      height: 100%;
      line-height: 28px;
      white-space: nowrap;
      text-overflow: ellipsis;
      overflow: hidden;
      padding: 0 16px;
      background-color: $white-color;
      color: $font-color;
      font-size: 14px;
      box-sizing: border-box;

      span {
        margin-left: 10px;
        font-size: 12px;
        color: $font3-color;
      }
    }
  }
}

.ta-sticky-view-item {
  &_header {
    width: 100%;
    height: 28px;
    line-height: 28px;
    white-space: nowrap;
    text-overflow: ellipsis;
    overflow: hidden;
    padding: 0 16px;
    background-color: $white-color;
    color: $font-color;
    font-size: 14px;
    box-sizing: border-box;

    span {
      margin-left: 10px;
      font-size: 12px;
      color: $font3-color;
    }

    .ta-sticky-view.disabled-header & {
      display: none;
    }
  }
}
