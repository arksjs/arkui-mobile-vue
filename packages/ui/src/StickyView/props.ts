import { isNumber, isString } from '../helpers/util'
import type { VoidFnToBooleanFn } from '../helpers/types'
import type { OnChange } from './types'

export const emitChangeValidator: VoidFnToBooleanFn<OnChange> = (
  name,
  activeIndex
) => isString(name) && isNumber(activeIndex)
