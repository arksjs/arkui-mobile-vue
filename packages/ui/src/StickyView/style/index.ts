import '../../style/common/style/index'
import '../../ScrollView/style/index'
import '../../Sticky/style/index'
import '../index.scss'
