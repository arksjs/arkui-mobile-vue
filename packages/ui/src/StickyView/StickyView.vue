<template>
  <div ref="root" :class="classes">
    <ScrollView
      class="ta-sticky-view_content"
      :enablePullDirections="enablePullDirections"
      :scrollY="isSelfContainer"
      :pullRefreshTexts="pullRefreshTexts"
      @refreshing="onPullRefreshing"
      ref="scrollViewRef"
    >
      <StickyViewList ref="itemsRef" @resetItems="resetItems">
        <slot></slot>
      </StickyViewList>
    </ScrollView>
    <Sticky
      v-show="!disabledHeader"
      :offsetTop="offsetTop"
      :containSelector="containSelector"
      :disabled="disabledHeader"
      class="ta-sticky-view_top"
      ref="stickyRef"
    >
      <div class="ta-sticky-view_fixed">
        <div class="ta-sticky-view_fixed-inner" ref="fixedEl"></div>
      </div>
    </Sticky>
  </div>
</template>

<script lang="ts">
import {
  computed,
  defineComponent,
  onMounted,
  ref,
  watch,
  nextTick,
  shallowRef,
  type PropType,
  type VNode
} from 'vue'
import { ScrollView } from '../ScrollView'
import { Sticky } from '../Sticky'
import {
  CSSProperties2CssText,
  getRelativeOffset,
  getScrollTop,
  getSizeValue,
  querySelector,
  scrollTo as _scrollTo,
  selectorValidator,
  isSizeValue,
  isNumber,
  isString,
  type PropsToEmits,
  type Selector,
  getElementItems
} from '../helpers'
import StickyViewList from './StickyViewList.vue'
import { useScroll, useException, useOnce } from '../hooks'
import { emitChangeValidator } from './props'
import type {
  StickyViewEmits,
  StickyViewItem,
  StickyViewListRef,
  StickyViewRef,
  StickyViewOnPullRefreshing,
  StickyViewPullRefreshTexts
} from './types'
import type { ResetContainer, StickyRef } from '../Sticky/types'
import { getClasses, getFixedStyles, FIXED_HEIGHT } from './util'
import type { ScrollViewRef } from '../ScrollView/types'
import { emitRefreshingValidator } from '../ScrollView/props'

export default defineComponent({
  name: 'ta-sticky-view',
  components: { Sticky, StickyViewList, ScrollView },
  props: {
    modelValue: {
      type: String
    },
    containSelector: {
      type: [String, HTMLElement] as PropType<Selector>,
      validator: selectorValidator
    },
    offsetTop: {
      type: [Number, String],
      validator: isSizeValue,
      default: 0
    },
    // 禁用吸附
    disabledHeader: {
      type: Boolean,
      default: false
    },
    // 允许上拉刷新
    enablePullRefreshUp: {
      type: Boolean,
      default: false
    },
    // 允许下拉刷新
    enablePullRefreshDown: {
      type: Boolean,
      default: false
    },
    // 下拉刷新提示文案
    pullRefreshTexts: {
      type: Object as PropType<StickyViewPullRefreshTexts>
    }
  },
  emits: {
    'update:modelValue': name => isString(name),
    change: emitChangeValidator,
    resetItems: items => {
      if (Array.isArray(items)) {
        return (
          items.filter(item => {
            return !(item && isNumber(item.index) && isString(item.name) && isString(item.title))
          }).length === 0
        )
      }
      return false
    },
    pullRefreshing: emitRefreshingValidator
  } as PropsToEmits<StickyViewEmits>,
  setup(props, { emit, expose }) {
    const { printListItemNotFoundError } = useException()
    const root = shallowRef<HTMLElement | null>(null)
    const container = shallowRef<HTMLElement | null>(null)
    const fixedEl = shallowRef<HTMLElement | null>(null)
    const stickyRef = shallowRef<StickyRef | null>(null)
    const scrollViewRef = shallowRef<ScrollViewRef | null>(null)
    const itemsRef = shallowRef<StickyViewListRef | null>(null)
    const activeIndex = ref(0)
    const isSelfContainer = ref(false)

    let cachedItems: StickyViewItem[] = []
    let $items: HTMLElement[] = []
    let isSpecifyScrolling = false // 是否指定滚动

    const once = useOnce()

    function getItemName(index: number) {
      return cachedItems[index]?.name || ''
    }

    function getItemTitle(index: number) {
      return (
        (cachedItems[index]?.title || '') + `<span>${cachedItems[index]?.description || ''}</span>`
      )
    }

    function getActiveIndexByName(name?: string) {
      if (name) {
        for (let i = 0; i < $items.length; i++) {
          if (getItemName(i) === name) {
            return i
          }
        }
      }

      return -1
    }

    let oldIndex = -1

    function onChange() {
      if (oldIndex !== activeIndex.value) {
        const name = getItemName(activeIndex.value)
        emit('update:modelValue', name)
        emit('change', name, activeIndex.value)
      }
      oldIndex = -1
    }

    function updateTitle(index: number, tY: number | null) {
      if (!fixedEl.value) {
        return
      }

      fixedEl.value.innerHTML = index === -1 ? '' : getItemTitle(index)
      fixedEl.value.style.cssText = CSSProperties2CssText(getFixedStyles(tY))
    }

    function updateFixed(ss: number | null) {
      if (!fixedEl.value || !container.value) {
        return
      }

      if ($items.length === 0) {
        updateTitle(-1, null)
        return
      }

      if (oldIndex === -1) {
        oldIndex = activeIndex.value
      }

      const scrollTop = ss == null ? getScrollTop(container.value) : ss

      const _index = activeIndex.value
      const nextIndex = _index + 1
      const offsetTops = getOffsetTops()
      const current = offsetTops[_index]
      const next = offsetTops[nextIndex] != null ? offsetTops[nextIndex] : Infinity
      const first = offsetTops[0]

      if (scrollTop < first) {
        updateTitle(-1, null)
      } else if (scrollTop >= current) {
        if (scrollTop >= next) {
          activeIndex.value = nextIndex
          updateTitle(nextIndex, 0)

          if (offsetTops[nextIndex + 1] && scrollTop >= offsetTops[nextIndex + 1]) {
            // 超过了
            updateFixed(scrollTop)
          } else if (!isSpecifyScrolling) {
            onChange()
          }
        } else if (next - scrollTop < FIXED_HEIGHT) {
          updateTitle(_index, next - scrollTop - FIXED_HEIGHT)
        } else {
          updateTitle(_index, 0)
        }
      } else {
        if (current - scrollTop < FIXED_HEIGHT) {
          updateTitle(_index - 1, current - scrollTop - FIXED_HEIGHT)
        } else {
          activeIndex.value = _index - 1
          updateTitle(_index - 1, 0)

          if (offsetTops[_index - 1] && offsetTops[_index - 1] > scrollTop) {
            updateFixed(scrollTop)
          } else if (!isSpecifyScrolling) {
            onChange()
          }
        }
      }

      isSpecifyScrolling &&
        once(() => {
          // 有一些指定滑动到相应位置，移动中间不需要不断上报onChange，只要上报最后一个
          isSpecifyScrolling = false
          onChange()
        })
    }

    function getOffsetTops() {
      const offset =
        getRelativeOffset(
          getListEl() as HTMLElement,
          (isSelfContainer.value ? root.value : container.value) as HTMLElement
        ).offsetTop - getSizeValue(props.offsetTop)

      return $items.map($el => {
        return $el.offsetTop + offset
      })
    }

    /**
     * 滚到到指定位置
     */
    function scrollToOffset(offset: number) {
      isSpecifyScrolling = true
      // 在onMounted后还需要nextTick才能有效调用滚动
      nextTick(() => _scrollTo(container.value as HTMLElement, offset, false))
    }

    /**
     * 滚动到第index个
     */
    function scrollToIndex(newIndex: number) {
      if ($items[newIndex]) {
        if (newIndex != activeIndex.value && container.value) {
          scrollToOffset(
            getRelativeOffset($items[newIndex], container.value).offsetTop -
              getSizeValue(props.offsetTop)
          )
        }
      } else {
        printListItemNotFoundError('index')
      }
    }

    /**
     * 滚动到指定name
     */
    function scrollTo(name: string) {
      const newIndex = getActiveIndexByName(name)

      if (newIndex !== -1) {
        scrollToIndex(newIndex)
      } else {
        printListItemNotFoundError('name')
      }
    }

    useScroll(container, () => updateFixed(null))

    const resetContainer: ResetContainer = containSelector => {
      const selfEl = scrollViewRef.value?.getScrollEl()

      if (!selfEl) {
        return
      }

      const newEl = querySelector(containSelector) || selfEl

      if (newEl === container.value) {
        return
      }

      container.value = newEl
      isSelfContainer.value = container.value === selfEl
      stickyRef.value?.resetContainer(newEl)

      updateFixed(null)
    }

    function getListEl() {
      return itemsRef.value?.ref || null
    }

    function isSameItems(a: StickyViewItem[], b: StickyViewItem[]) {
      if (a.length !== b.length) return false
      for (let i = 0; i < a.length; i++) {
        if (!(a[i].name === b[i].name && a[i].title === b[i].title)) {
          return false
        }
      }
      return true
    }

    function resetItems(_items: VNode[]) {
      $items = getElementItems(getListEl(), 'ta-sticky-view-item')

      updateFixed(null)

      const newItems: StickyViewItem[] = _items.map((item, index) => {
        return {
          index,
          name: item.props?.name || '',
          title: item.props?.title || item.props?.name || '',
          description: item.props?.description || ''
        }
      })

      if (!isSameItems(newItems, cachedItems)) {
        cachedItems = newItems
        emit('resetItems', newItems)
      }
    }

    function updateValue(val?: string) {
      const newIndex = getActiveIndexByName(val)

      if (newIndex !== -1) {
        if (newIndex != activeIndex.value) {
          // 把oldIndex设置为最新，阻止onChange被调用
          oldIndex = newIndex
          scrollToIndex(newIndex)
        }
      } else {
        printListItemNotFoundError('modelValue', true)
      }
    }

    watch(() => props.modelValue, updateValue)

    const enablePullDirections = computed(() => {
      const directions: ('up' | 'down')[] = []
      if (isSelfContainer.value) {
        if (props.enablePullRefreshUp) {
          directions.push('up')
        }
        if (props.enablePullRefreshDown) {
          directions.push('down')
        }
      }
      return directions
    })

    const onPullRefreshing: StickyViewOnPullRefreshing = (payload, loadComplete) => {
      emit('pullRefreshing', payload, loadComplete)
    }

    onMounted(() => {
      resetContainer(props.containSelector)
      $items = getElementItems(getListEl(), 'ta-sticky-view-item')
      // 首次设置 modelValue 非string类型认为是没配置，不做更新
      props.modelValue != null && updateValue(props.modelValue)
    })

    const classes = computed(() => getClasses(isSelfContainer.value, props.disabledHeader))

    expose({
      scrollTo,
      scrollToIndex,
      scrollToOffset,
      resetContainer
    } as StickyViewRef)

    return {
      root,
      isSelfContainer,
      fixedEl,
      itemsRef,
      stickyRef,
      scrollViewRef,
      classes,
      resetItems,
      enablePullDirections,
      onPullRefreshing,

      scrollTo,
      scrollToIndex,
      scrollToOffset,
      resetContainer
    }
  }
})
</script>
