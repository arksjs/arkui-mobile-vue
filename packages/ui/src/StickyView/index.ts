import { multiWithInstall } from '../helpers/with-install'
import StickyView from './StickyView.vue'
import StickyViewItem from './StickyViewItem.vue'

export { StickyView, StickyViewItem }
export default StickyView
export const install = multiWithInstall(StickyView, [StickyViewItem])
