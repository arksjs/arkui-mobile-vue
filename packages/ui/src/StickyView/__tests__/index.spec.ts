import { describe, test, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import { StickyView, StickyViewItem } from '@/StickyView'

describe('StickyView', () => {
  test('snapshot', () => {
    const wrapper = mount(StickyView, {})

    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('StickyViewItem', () => {
  test('snapshot', () => {
    const wrapper = mount(StickyViewItem, {
      props: {
        name: 'name'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
