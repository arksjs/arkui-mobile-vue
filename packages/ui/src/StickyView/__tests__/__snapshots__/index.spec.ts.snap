// Vitest Snapshot v1

exports[`StickyView > snapshot 1`] = `
"<div class=\\"ta-sticky-view\\">
  <div class=\\"ta-sticky-view_list\\"></div>
  <div class=\\"ta-sticky ta-sticky-view_top\\">
    <div class=\\"ta-sticky_content\\">
      <div class=\\"ta-sticky-view_fixed\\">
        <div class=\\"ta-sticky-view_fixed-inner\\" style=\\"transform: translate3d(0px, -100%, 0px);\\"></div>
      </div>
    </div>
  </div>
</div>"
`;

exports[`StickyViewItem > snapshot 1`] = `
"<div class=\\"ta-sticky-view-item\\" data-name=\\"name\\" data-title=\\"\\">
  <div class=\\"ta-sticky-view-item_header\\">name</div>
  <div class=\\"ta-sticky-view-item_body\\"></div>
</div>"
`;
